### PR TITLE
Propose OpenTelemetry Profiling Vision

### PR DESCRIPTION
**Pull request by Rperry2174**
This change proposes high-level items that define our long-term vision for
Profiling support in OpenTelemetry project. 

A group of open source maintainers, vendors, end-users, and developers excited about profiling / otel have been meeting for ~8 weeks now and this document was created collectively in order to share with the broader community for feedback. We've had ~15 people contribute directly to the document and over 60 people who have attended our meetings over the past couple of months. 

This idea of "Adding profiling as a support event type" has also been discussed at length in [this issue](https://github.com/open-telemetry/oteps/issues/139) created in October 2020. 

If this proposal is accepted/approved then we will proceed with filling out this [project tracking issue](https://github.com/open-telemetry/opentelemetry-specification/issues/2731) and following the other procedures outlined in the [project management instructions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/project-management.md). 

Comments, ideas, feedback, etc. are all very welcome and highly appreciated!

### CODE REVIEW

File: text/profiles/0212-profiling-vision.md
Comment (inge4pres) [inline text/profiles/0212-profiling-vision.md:93]: grammar nit: `alotment` -> `allotment`

File: text/profiles/0212-profiling-vision.md
Comment (tigrannajaryan) [inline text/profiles/0212-profiling-vision.md:1]: Editorial: you may want to hard wrap the lines to some reasonable number f characters (e.g. 80 or 100) to make reviewing easier. Otherwise comments are attached to entire paragraphs instead of targeting individual lines where necessary.

File: text/profiles/0212-profiling-vision.md
Comment (tigrannajaryan) [inline text/profiles/0212-profiling-vision.md:51]: >The nature of profiling also provides the potential for a more “hands-off” approach to adding it to code compared to manual instrumentation needed for other signals.

I am not sure why this sentence is added to this particular section. It seems irrelevant and the comparison is also unclear. Why compare with manual instrumentation and not to autoinstrumentation of other signals which is also often hands-off?

File: text/profiles/0212-profiling-vision.md
Comment (tigrannajaryan) [inline text/profiles/0212-profiling-vision.md:65]: What is "transcoding" in this context? Is this about translation to/from other profiling formats?

File: text/profiles/0212-profiling-vision.md
Comment (tigrannajaryan) [inline text/profiles/0212-profiling-vision.md:42]: This is a great link, thank you.

File: text/profiles/0212-profiling-vision.md
Comment (tigrannajaryan) [inline text/profiles/0212-profiling-vision.md:67]: Editorial: this one is a bit difficult to understand. It is unlike the previous 3 bullet points, which list what "should be". Can you make this clearer ([this doc about parallel vs nonparallel lists](https://developers.google.com/tech-writing/one/lists-and-tables#keep_list_items_parallel) may be useful)?

File: text/profiles/0212-profiling-vision.md
Comment (Rperry2174) [reply text/profiles/0212-profiling-vision.md:51]: yeah true. The idea here was to say:
1. Linking tracing to other signals is very useful
2. Linking tracingto other signals is _easy_ (since it isn't a very manual process) 

Don't feel super strongly about that second part though we could reword or remove

File: text/profiles/0212-profiling-vision.md
Comment (Rperry2174) [reply text/profiles/0212-profiling-vision.md:65]: yes -- do you think it's too ambiguous without saying that explicitly?

File: text/profiles/0212-profiling-vision.md
Comment (tigrannajaryan) [reply text/profiles/0212-profiling-vision.md]: It it is not entirely true that manual instrumentation is needed for other signals to capture trace context. For example logging SDKs in some languages can also automatically capture the current trace context and emit trace id/span id in the logs without any manual instrumentation needed.

>Linking to other signals is easy

It is probably useful to say this ^^^ directly instead of bringing the argument that it is _easier_ for profiling than for other signals.

File: text/profiles/0212-profiling-vision.md
Comment (tigrannajaryan) [reply text/profiles/0212-profiling-vision.md]: Yes, if possible say that explicitly since it wasn't entirely clear to me.

**Review by tigrannajaryan — APPROVED** [review]
Mostly LGTM, thank you.

Approving, this is aligned with my vision as well.

File: text/profiles/0212-profiling-vision.md
Comment (lizthegrey) [inline text/profiles/0212-profiling-vision.md:174]: define "code-level" granularity e.g. "function call and line of code"

**Review by lizthegrey — APPROVED** [review]

File: text/profiles/0212-profiling-vision.md
Comment (Rperry2174) [reply text/profiles/0212-profiling-vision.md]: We go more into the details about the process of linking to other signals later so I'll just remove this part

File: text/profiles/0212-profiling-vision.md
Comment (reyang) [inline text/profiles/0212-profiling-vision.md:166]:
```suggestion
- Understanding what code is responsible for consuming resources (i.e. CPU, RAM,
```

File: text/profiles/0212-profiling-vision.md
Comment (reyang) [inline text/profiles/0212-profiling-vision.md:53]:
```suggestion
  time-to-value by often being able to optionally drop in a minimal amount of
```

File: text/profiles/0212-profiling-vision.md
Comment (reyang) [inline text/profiles/0212-profiling-vision.md:56]:
```suggestion
  across different languages, but with a little effort the representation of
```

**Review by reyang — APPROVED** [review]

**Comment by halcyondude** [comment]
+1 NB

**Review by jsuereth — APPROVED** [review]

File: text/profiles/0212-profiling-vision.md
Comment (SeanHeelan) [inline text/profiles/0212-profiling-vision.md:135]:
```suggestion
standardization effort it is not feasible to be prescriptive on the matter of
```

Typo fix. When I wrote this I mean prescriptive, not proscriptive.

**Review by SeanHeelan — APPROVED** [review]
lgtm, just a couple of minor wording fixes from the performance section I added in the original doc.

**Review by carlosalberto — APPROVED** [review]

**Review by aalexand — APPROVED** [review]

**Review by petethepig — APPROVED** [review]

File: text/profiles/0212-profiling-vision.md
Comment (aalexand) [inline text/profiles/0212-profiling-vision.md:89]: nit: make all headings use sentence case.

https://docs.microsoft.com/en-us/style-guide/capitalization

**Review by djaglowski — APPROVED** [review]

**Review by kolesnikovae — APPROVED** [review]

File: text/profiles/0212-profiling-vision.md
Comment (dgolant) [inline text/profiles/0212-profiling-vision.md:169]: I would generalize this to something like "Tracking resource utilization of an application over time to understand how code changes, hardware configuration changes, and ephemeral environmental issues influence performance" or something along those lines. In my experience the observations that profiling can yield that are most useful are the ones divorced from code changes.

**Review by dgolant — APPROVED** [review]

**Review by justinabrahms — APPROVED** [review]

**Comment by yurishkuro** [comment]
I am supportive of this vision, but there seems to be a glaring omission in the document - the terms "profile" and "profiling" are never defined. Since this is essentially proposing a new type of telemetry signal to be included in the project, can you provide a clear definition?

File: text/profiles/0212-profiling-vision.md
Comment (crobert-1) [inline text/profiles/0212-profiling-vision.md:103]: Nit: capitalization should be consistent (probably lowercase) with correlation type names (trace context correlation vs Resource Context Correlation).

File: text/profiles/0212-profiling-vision.md
Comment (crobert-1) [inline text/profiles/0212-profiling-vision.md:160]: Tiny nit: This document goes has both "Otel" and "OTEL" as abbreviations for OpenTelemetry. It should probably be defined as an abbreviation early on, and then stick to one casing.

**Review by crobert-1 — APPROVED** [review]
Just some minor nits, otherwise LGTM!

File: text/profiles/0212-profiling-vision.md
Comment (yurishkuro) [reply text/profiles/0212-profiling-vision.md:160]: better yet, search/replace OTEL->OpenTelemetry. Abbreviations are good for twitter and slack, in the official documents like OTEPs a full name is preferred.

**Comment by halcyondude** [comment]
This captures the vision and spirit of the discussions over the past few months. I support!

+1

**Review by Kielek — APPROVED** [review]

**Review by florianl — APPROVED** [review]

File: text/profiles/0212-profiling-vision.md
Comment (kittylyst) [inline text/profiles/0212-profiling-vision.md:38]: Consider: "consistent instrumentation based upon profiling data."

File: text/profiles/0212-profiling-vision.md
Comment (kittylyst) [inline text/profiles/0212-profiling-vision.md:49]: As it stands, we are vague about what type(s) of profiling we mean - and it only becomes more concrete at the end of the doc that we are talking about not only Execution Profiling but Memory, Disk, etc.

I think it would flow better if that ambiguity was resolved up front - if we make it concrete that multiple types of profiling are in-scope.

File: text/profiles/0212-profiling-vision.md
Comment (kittylyst) [inline text/profiles/0212-profiling-vision.md:149]: It seems that we're glossing over the always-on / activated-on-demand question here. Is that the intent? Do we feel that the standard should support both?

File: text/profiles/0212-profiling-vision.md
Comment (kittylyst) [reply text/profiles/0212-profiling-vision.md:169]: +1

**Review by kittylyst — COMMENTED** [review]
Decent statement of high-level intent.

**Review by scbjans — APPROVED** [review]

File: text/profiles/0212-profiling-vision.md
Comment (Rperry2174) [reply text/profiles/0212-profiling-vision.md:1]: yeah I've been thinking about this and not sure how "deep" to go for this definition. Do you know of an example I can look to for i.e. metrics or tracing particularly in the context of otel?

Comment (yurishkuro) [reply text/profiles/0212-profiling-vision.md:1]: Fair point, I also don't see definition of metrics / traces, so I opened respective tickets.

There is a group of definitions related to Logs - https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#logs

File: text/profiles/0212-profiling-vision.md
Comment (Rperry2174) [reply text/profiles/0212-profiling-vision.md:149]: Yeah in this case we are intentionally leaving this vague as we will further dig into the details of this in a future OTEP

**Comment by chadbrewbaker** [comment]
If I had to make it more concrete - I might adopt the Rust targets for OS/libc/CPU as a first approximation - especially for embedded CPUs.

On a geospatial level - lat/long would be useful data for network flow modeling. Also network triangulation like pings to three servers - bandwidth might be harder to measure.

File: text/profiles/0212-profiling-vision.md
Comment (Rperry2174) [reply text/profiles/0212-profiling-vision.md]: Added in definitions as discussed with the rest of the working group. Let me know if you have any other feedback @yurishkuro

**Comment by Rperry2174** [comment]
I believe all comments have been addressed let me know if anything else needs to be before merging!

**Review by yurishkuro — APPROVED** [review]

File: text/profiles/0212-profiling-vision.md
Comment (Kielek) [inline text/profiles/0212-profiling-vision.md]:
```suggestion
depending on the context, for the purposes of this OTEP we are defining the two
```

**Review by dashpole — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
This vision now has a large number of approvals from existing OpenTelemetry members, from OpenTelemetry Technical Committee and Governance Committee and from the new Profiling Workgroup members.

I believe this shows a clear alignment on the vision from all participants.

Members of @open-telemetry/technical-committee if you haven't had a chance to review this vision, please do. 

I will merge the PR and will consider this current vision accepted by OpenTelemetry Profiling if I don't see any further comments.

**Review by kittylyst — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
No new comments. Lots of approvals. Merging.

**Pull request merged**
